src/envoy/http/backend_routing/filter.cc

```diff
@@ -52,50 +54,76 @@ FilterHeadersStatus Filter::decodeHeaders(
-        utils::getStringFilterState(filter_state, utils::kQueryParams);
-    const auto originalPath = std::string(original_path);
-    std::size_t originalQueryParamPos = originalPath.find('?');
-    if (originalQueryParamPos != std::string::npos) {
-      // has query parameters in original url
-      const std::string& originalQueryParam =
-          originalPath.substr(originalQueryParamPos);
-      newPath = absl::StrCat(rule->path_prefix(), originalQueryParam);
-      if (!queryParamFromPathParam.empty()) {
-        absl::StrAppend(&newPath, "&", queryParamFromPathParam);
-      }
-    } else {
-      newPath = rule->path_prefix();
-      if (!queryParamFromPathParam.empty()) {
-        absl::StrAppend(&newPath, "?", queryParamFromPathParam);
-      }
-    }
-    config_->stats().constant_address_request_.inc();
-    ENVOY_LOG(debug,
-              "constant address backend routing for operation {}, new path: {}",
-              operation, newPath);
-  } else {  // APPEND_PATH_TO_ADDRESS
-    newPath = absl::StrCat(rule->path_prefix(), original_path);
-    config_->stats().append_path_to_address_request_.inc();
-    ENVOY_LOG(
-        debug,
-        "append path to address backend routing for operation {}, new path: {}",
-        operation, newPath);
+
+  absl::string_view original_path = headers.Path()->value().getStringView();
+  std::string new_path;
+
+  switch (rule->path_translation()) {
+    case BackendRoutingRule::CONSTANT_ADDRESS:
+      new_path = translateConstPath(rule->path_prefix(), original_path);
+      config_->stats().constant_address_request_.inc();
+      ENVOY_LOG(debug,
+                "constant address backend routing for operation {}"
+                ", original path: {}, new path: {}",
+                operation, original_path, new_path);
+      break;
+
+    case BackendRoutingRule::APPEND_PATH_TO_ADDRESS:
+      new_path = translateAppendPath(rule->path_prefix(), original_path);
+      config_->stats().append_path_to_address_request_.inc();
+      ENVOY_LOG(debug,
+                "append path to address backend routing for operation {}"
+                ", original path: {}, new path: {}",
+                operation, original_path, new_path);
+      break;
+
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
   }
-  const auto& pathField = Envoy::Http::Headers::get().Path;
-  headers.remove(pathField);
-  headers.addCopy(pathField, newPath);
 
+  headers.setPath(new_path);
   return FilterHeadersStatus::Continue;
 }
 
+// Replace the original path with the constant path prefix.
+// If any variable bindings were extracted, append them as query params.
+std::string Filter::translateConstPath(absl::string_view prefix,
+                                       absl::string_view original_path) {
+  const auto& filter_state = *decoder_callbacks_->streamInfo().filterState();
+  absl::string_view extracted_query_params =
+      utils::getStringFilterState(filter_state, utils::kQueryParams);
```

Comment:
Since this is the only place to use class member, can we inject extracted_query_params, so that move `translateConstPath` and `translateAppendPath` functions out of class interface to keep it simple? Just my 2 cents, what do you think?@qiwzhang

Reply:
Is there any specific benefit to moving these to an anonymous namespace? If not, I feel it's fine to keep them as private functions in the class. Also it helps with readability to have all the functions declared in the header file. Up to you

Reply:
private member function is fine as I see reading FilterState is specific to ConstPath,  it should be here at this function.